### ISSUE TITLE
[Bug]: Compressed chunks are heavier than uncompressed ones (by far)

### ISSUE THREAD

**Issue by mickael-choisnet** [timescale/timescaledb#5769]
### What type of bug is this?

Performance issue

### What subsystems and features are affected?

Compression

### What happened?

I wanted to test the new features regarding the compressed chunks with the latest version (2.11.0).
I updated the extension on my existing cluster (update from 2.10.1) and created a new table with some fake data to try and comprehend the new mechanic and what happens behind the scene.
My test stopped when I received a message telling me I had no left space on my disk during compression. 
I was compressing my chunks, and the initial volume of ~38 GB has grown up to take all the remaining 70 GB space left on my disk.

This feels like a huge step back from the compression performance we had with the 2.10.X and prior versions.
Is there a specific reason for this? Is it by design? (I hope not, given the ratio of x4 or x5 on the initial size)
Does someone else reported this yet, or am I the only one to have this change of behavior?

### TimescaleDB version affected

2.11.0

### PostgreSQL version used

PostgreSQL 14.5 (Debian 14.5-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit

### What operating system did you use?

Debian GNU/Linux 10 (buster) (10.13)

### What installation method did you use?

Deb/Apt

### What platform did you run on?

Microsoft Azure Cloud

### How can we reproduce the bug?

1. create a new database with timescaledb 2.11.0

   ```sql
   CREATE DATABASE test-db;
   CREATE EXTENSION timescaledb VERSION '2.11.0';
   ```

2. create tables

   ```sql
   CREATE TABLE public.measures (
       measure_date TIMESTAMP NOT NULL,
       received_date TIMESTAMP NOT NULL,
       value DOUBLE PRECISION,
       meter_id TEXT NOT NULL,
       fluid_id SMALLINT NOT NULL,
       source SMALLINT NOT NULL,
       device_type SMALLINT NULL,
       device_id TEXT NULL
   );
   SELECT create_hypertable('measures', 'measure_date', chunk_time_interval => INTERVAL '5 DAYS', create_default_indexes => false);
   ALTER TABLE measures SET (
       timescaledb.compress,
       timescaledb.compress_segmentby = 'meter_id, fluid_id'
   );
   SELECT add_compression_policy(hypertable => 'measures', compress_after => INTERVAL '14 DAYS', if_not_exists => true);
   ```

3. create proc to generate measures

   ```sql
   CREATE OR REPLACE PROCEDURE public.load_measures (IN date_start TIMESTAMP, IN date_end TIMESTAMP, IN gs_start INTEGER, IN gs_end INTEGER) AS
   $BODY$
   DECLARE
       inserted_at TIMESTAMP;
   BEGIN
       SELECT current_timestamp::TIMESTAMP INTO inserted_at;
   
       WITH RECURSIVE some_days AS (
           SELECT date_start AS rdate 
           UNION ALL 
           SELECT rdate + INTERVAL '1 DAY'
           FROM some_days
           WHERE rdate < date_end
       )
       INSERT INTO measures(measure_date, received_date, value, meter_id, fluid_id, source, device_type, device_id)
       SELECT
           d.rdate + (floor(random() * 22) || ' HOURS')::INTERVAL AS measure_date, 
           inserted_at AS received_date, 
           floor(random() * 10000 + 1000) AS value, 
           CAST(gs.id as TEXT) AS meter_id, 
           floor(random() * 8 + 1) AS fluid_id, 
           floor(random() * 10 + 1) AS source, 
           floor(random() * 8 + 1) AS device_type, 
           'ABCDEF' AS device_id
       FROM (SELECT generate_series(gs_start, gs_end) AS id) gs
       INNER JOIN lateral (SELECT * FROM some_days) d ON true;
   END;
   $BODY$ 
   LANGUAGE PLPGSQL;
   ```

4. load some data

   ```sql
   CALL load_measures(date_start => '2023-03-01', date_end => '2023-05-01', gs_start => 1000000, gs_end => 1500000);
   ```
   
5. trigger the compression job or manually crompress the chunks

   ```sql
   DO
   $BODY$
       DECLARE
           job_id  INT;
       BEGIN
           SELECT job.id INTO job_id
           FROM _timescaledb_config.bgw_job job
           INNER JOIN _timescaledb_catalog.hypertable hypertable ON hypertable.id = job.hypertable_id
           WHERE hypertable.table_name = 'measures';
   
           PERFORM alter_job(job_id, next_start=> current_timestamp);
       END;
   $BODY$;
   
   -- OR --
      
   DO
   $BODY$
       DECLARE
           chunk   RECORD;
       BEGIN
           FOR chunk IN SELECT show_chunks('measures', older_than => ('14 DAYS')::INTERVAL) AS name
           LOOP
               RAISE NOTICE '% - Compressing chunk "%"...', timeofday(), chunk.name;  
               PERFORM compress_chunk(chunk.name, if_not_compressed => true);
               COMMIT;
           END LOOP;
       END;
   $BODY$;
   ```

6. query usefull data

   ```sql
      WITH chunks_tr AS (
       SELECT
           format('%1$I.%2$I', ch.schema_name, ch.table_name)::TEXT AS chunk,
           tsrange(
               CASE WHEN sl.range_start = -9223372036854775808 THEN NULL
                   ELSE _timescaledb_internal.to_timestamp_without_timezone(sl.range_start) END,
               CASE WHEN sl.range_end = 9223372036854775807 THEN NULL
                   ELSE _timescaledb_internal.to_timestamp_without_timezone(sl.range_end) END
           ) AS time_range
       FROM _timescaledb_catalog.chunk ch
       JOIN _timescaledb_catalog.hypertable ht ON ch.hypertable_id = ht.id
       JOIN _timescaledb_catalog.dimension di ON di.hypertable_id = ht.id
       JOIN _timescaledb_catalog.chunk_constraint cn ON cn.chunk_id = ch.id
       JOIN _timescaledb_catalog.dimension_slice sl ON cn.dimension_slice_id = sl.id
       WHERE di.column_type = 'timestamp'::regtype
   ),
   chunk_relation AS (
       SELECT
           cds.chunk_schema as schema_name,
           cds.chunk_name as chunk_name,
           chunks_tr.time_range as ranges,
           cds.table_bytes as heap, 
           cds.index_bytes as index, 
           cds.toast_bytes as toast, 
           cds.total_bytes as total
       FROM chunks_detailed_size('measures') cds
       JOIN chunks_tr ON chunks_tr.chunk = cds.chunk_schema || '.' || cds.chunk_name
   )
   SELECT
       cr.schema_name || '.' || cr.chunk_name::text as chunk_name, 
       compr.table_name::text as compr_chunk_name, 
       cr.ranges, 
       cs.compression_status as status,
       CASE WHEN cs.uncompressed_heap_size IS NULL THEN cr.heap ELSE cs.uncompressed_heap_size END /1024/1024 || ' MB' as heap,
       CASE WHEN cs.uncompressed_index_size IS NULL THEN cr.index ELSE cs.uncompressed_index_size END /1024/1024 || ' MB' as index,
       CASE WHEN cs.uncompressed_toast_size IS NULL THEN cr.toast ELSE cs.uncompressed_toast_size END /1024/1024 || ' MB' as toast,
       CASE WHEN cs.uncompressed_total_size IS NULL THEN cr.total ELSE cs.uncompressed_total_size END /1024/1024 || ' MB' as total,
       cs.compressed_heap_size/1024/1024 || ' MB' as heap_compress, 
       cs.compressed_index_size/1024/1024 || ' MB' as index_compress, 
       cs.compressed_toast_size/1024/1024 || ' MB' as toast_compress, 
       cs.compressed_total_size/1024/1024 || ' MB' as total_compress
   FROM chunk_relation cr
   JOIN _timescaledb_catalog.chunk hyper ON hyper.table_name = cr.chunk_name
   LEFT JOIN _timescaledb_internal.compressed_chunk_stats cs ON cr.chunk_name::text = cs.chunk_name::text
   LEFT JOIN _timescaledb_catalog.chunk compr ON compr.id = hyper.compressed_chunk_id
   ORDER by cr.ranges DESC
   ```

**Comment by sb230132** [comment]
Hi @mickael-choisnet 
Thanks for the detailed description of the problem. In brief you are 
creating a hypertable,  enabling compression, loading data and then compressing all chunks.

New features introduced in 2.11.0 are related to performing INSERT/UPDATE/DELETE on compressed chunks. I dont see any of these operations being done in the repro steps.

However i can see that compression ratio is very badly affected. Will look into it to understand where is the issue and get back to you.

**Comment by mickael-choisnet** [comment]
Hi @sb230132,

Thank you for your fast answer and your analysis.
In fact, I did not report any misbehavior regarding the new features. I haven't had time to check this yet, I was just giving my test case some context.
Indeed, I was a bit harsh with my test and did not see that I was doing a poor job creating so many unique "key" values, sorry about that.

So I did more tests this morning, between timescaledb 2.11.0 and 2.9.3 (we did not deploy 2.10.x on our PROD due to a segfault error we encountered while testing it).
To begin with, I changed my Hypertable structure to allow a greater range (14 DAYS as we have in our production) with a compression after 30 DAYS.

   ```sql
   CREATE TABLE public.measures (
       measure_date TIMESTAMP NOT NULL,
       received_date TIMESTAMP NOT NULL,
       value DOUBLE PRECISION,
       meter_id TEXT NOT NULL,
       fluid_id SMALLINT NOT NULL,
       source SMALLINT NOT NULL,
       device_type SMALLINT NULL,
       device_id TEXT NULL
   );
   SELECT create_hypertable('measures', 'measure_date', chunk_time_interval => INTERVAL '14 DAYS', create_default_indexes => false);
   ALTER TABLE measures SET (
       timescaledb.compress,
       timescaledb.compress_segmentby = 'meter_id, fluid_id'
   );
   SELECT add_compression_policy(hypertable => 'measures', compress_after => INTERVAL '30 DAYS', if_not_exists => true);
   ```

I also updated the randomization of the fluid_id to make it static (which it is anyway, far more close to our real world data).

   ```sql
   CREATE OR REPLACE PROCEDURE public.load_measures (IN date_start TIMESTAMP, IN date_end TIMESTAMP, IN gs_start INTEGER, IN gs_end INTEGER) AS
   $BODY$
   DECLARE
       inserted_at TIMESTAMP;
   BEGIN
       SELECT current_timestamp::TIMESTAMP INTO inserted_at;
   
       WITH RECURSIVE some_days AS (
           SELECT date_start AS rdate 
           UNION ALL 
           SELECT rdate + INTERVAL '1 DAY'
           FROM some_days
           WHERE rdate < date_end
       )
       INSERT INTO measures(measure_date, received_date, value, meter_id, fluid_id, source, device_type, device_id)
       SELECT
           d.rdate + (floor(random() * 22) || ' HOURS')::INTERVAL AS measure_date, 
           inserted_at AS received_date, 
           floor(random() * 10000 + 1000) AS value, 
           CAST(gs.id as TEXT) AS meter_id, 
           1 AS fluid_id, 
           floor(random() * 10 + 1) AS source, 
           floor(random() * 8 + 1) AS device_type, 
           'ABCDEF' AS device_id
       FROM (SELECT generate_series(gs_start, gs_end) AS id) gs
       INNER JOIN lateral (SELECT * FROM some_days) d ON true;
   END;
   $BODY$ 
   LANGUAGE PLPGSQL;
   ```

With all these adjustments, I ran my scenario on both databases (one with 2.11.0 and one with 2.9.3) and I can relate that the compression ratio is still the same.
I also ran my initial case (with smaller chunk interval and more unique data) and it was also a poor compression result on 2.9.3.

---
However, one thing worries me a bit. 
If the timespan is too small for the chunks (7 days for example), even with many rows per unique SEGMENTED BY values, the compression ratio starts to go the other way around.

With chunks of 14 days timespan:

- distinct `meter_id/fluid_id` by 'full' chunk: 1 000 000
- total `row count` by 'full' chunk: 14 000 000
- compression ratio is good:
![image](https://github.com/timescale/timescaledb/assets/13468399/5c74115f-a035-4703-b298-7a6487157243)

With chunks of 7 days timespan:

- distinct `meter_id/fluid_id`by 'full' chunk: 1 000 000
- total `row count` by 'full' chunk: 7 000 000
- compression ratio starts to get very low:
![image](https://github.com/timescale/timescaledb/assets/13468399/357e942f-209d-4cba-81bf-1851da28ef2b)

So, if we continue to get more and more new devices and our chunks starts to grow out of the recommended range (1/4 of the available RAM for all last chunks of all Hypertables), we will not be able to play long with the chunk timespan to adjust its size and will soon have to consider upgrading our virtual machine to have more RAM.
Am I correct in this analyze?

I find the point of inversion (effective compression vs more space occupied after compression) to happen way too soon.
7 rows per unique ID seems a reasonable enough ratio to get a space downsize after compression...
I'll grant you that with 7 days timespan there is a gain on the heap space and the downsize is due to the creation of the index. But it seems to be the limit before we get a poor compression ratio.

Is that a normal behavior? Is this 'point of inversion' a subject studied for future improvement?

**Comment by mickael-choisnet** [comment]
I've done the test with a timespan of 6 days per chunk to be sure:

- distinct `meter_id/fluid_id` by 'full' chunk: 1 000 000
- total `row count` by 'full' chunk: 14 000 000

![image](https://github.com/timescale/timescaledb/assets/13468399/a3bbc85f-0abc-4f19-8861-07df2d5b4e11)

The threshold is indeed 7 rows per unique key in a chunk before the compression rate starts to be > 1.

**Comment by sb230132** [comment]
Hi @mickael-choisnet 

In short you are right. However an improvement can be done by defining ORDER BY.
As per https://docs.timescale.com/self-hosted/latest/compression/#order-entries if you specify ORDER BY settings you may see improvements in compression ratio.

**Comment by mickael-choisnet** [comment]
Thank you, @sb230132, for the information, I have run a test with the "6 days timespan case", and it is impressive (`timescaledb.compress_orderby = 'meter_id, fluid_id, measure_date DESC'` and no `compress_segmentby` option):
![image](https://github.com/timescale/timescaledb/assets/13468399/c453be70-1187-4037-b735-6211be90867e)

But I wonder... Does this will affect my queries upon the compressed chunks? 
Does my queries targeting principally `meter_id` , `fluid_id` and `measure_date` will still be as efficient as before with this new organization?

It seems this new configuration does not involve the creation of an index for the compressed chunk and uses the toast (which is 
something I know nothing about...)

**Comment by sb230132** [comment]
@mickael-choisnet 
If there is no `compress_segmentby` defined, then by default time column (2nd parameter) specified in create_hypertable() will be used to group the rows before compression. Since `measure_date` column values are not so skewed you see the above results.

Queries on compressed chunks using `meter_id` , `fluid_id` and `measure_date` may improve because now data in each compressed rows are ordered based on these columns.

Iam not sure what kind of queries are run on compressed chunks. Please share those queries so that i can confirm and let you know if queries are affected with this change in compression settings.

For example:
SELECT * FROM measures  WHERE measure_date = '2023-04-11 12:23:23'; will be more efficient than
SELECT * FROM measures  WHERE fluid_id = 1 AND meter_id = '1000000';

**Comment by mickael-choisnet** [comment]
Hi @sb230132,

Ok, thanks for the explanation about the compression and the difference with or without `segmentby option`.
I think I will run some tests anyway to find out if this way of compressing our chunks could prove more efficient!

To answer your request, here is an example query we use:
```sql
SELECT * -- OR SOME AGGREGATE FUNC
FROM measures i
-- This LEFT JOIN is specific to one of our Hypertable. measures_realtime is a Hypertable with only 2 days of data and is light weight
LEFT JOIN measures_realtime ir ON ir.meter_id = i.meter_id AND ir.fluid_id  = ir.fluid_id AND ir.measure_date = i.measure_date
INNER JOIN meters m ON m.id = i.meter_id
INNER JOIN room r ON r.id = m.room_id
INNER JOIN area a ON a.id = r.area_id
INNER JOIN jsonb_array_elements(@AreaFluidFilters) arr(af) ON arr.af->>'area_code' = area.code AND (arr.af->>'fluid_id')::smallint = i.fluid_id
WHERE ir.appareil_id IS NULL
    -- AND r.origine_id = @RoomId -- this can be use instead of the INNER JOIN jsonb_array_elements sometimes
    AND i.measure_date >= @StartDate AND i.measure_date <= @EndDate
    AND i.fluid_id = @FluidId
```

Assuming the other tables are common tables with the right indexes defined but without any FK. 
These table's only purpose is to be able to query our Hypertables.

In other words, we always apply a filter to the time dimension, but it is mostly a timespan (never a specific timestamp) and it is always paired with a join on some of our metadata tables to filter on some specific scope.

**Comment by sb230132** [comment]
Hi @mickael-choisnet 
As long as filters are based on time dimension, the new compression settings should not affect the queries.

**Comment by mickael-choisnet** [comment]
OK !

I'm closing this issue (which was not a bug after all).
Thank you for your time and your answers @sb230132

**Comment by sb230132** [comment]
hi @mickael-choisnet 
In this comment 
https://github.com/timescale/timescaledb/issues/5769#issuecomment-1584302649
you mentioned about segfault.
Can you please give me more details about this along with repro steps.

**Comment by mickael-choisnet** [comment]
Yes, I created another issue about this, and it has been resolved as far as I can tell (but I haven't had time to check that on my own)

Here is the related bug I created at the time :  #5434
